In [9]:
import torch
import torch.distributions as dist

In [47]:
def p_C(c:torch.Tensor)->torch.Tensor:
    probs = torch.tensor([0.5,0.5])
    d = dist.Categorical(probs)
    return torch.exp(d.log_prob(c))


def p_S_given_C(s: torch.Tensor,c: torch.Tensor) -> torch.Tensor:
    probs = torch.tensor([[0.5, 0.9], [0.5, 0.1]])
    d = dist.Categorical(probs.t())
    lp = d.log_prob(s)[c.detach()]
    return torch.exp(lp)


def p_R_given_C(r: torch.Tensor, c: torch.Tensor) -> torch.Tensor:
    probs = torch.tensor([[0.8, 0.2], [0.2, 0.8]])
    d = dist.Categorical(probs.t())
    lp = d.log_prob(r)[c.detach()]
    return torch.exp(lp)

def p_W_given_S_R(w: torch.Tensor, s: torch.Tensor, r: torch.Tensor) -> torch.Tensor:
    probs = torch.tensor([
        [[1.0, 0.1], [0.1, 0.01]],  # w = False
        [[0.0, 0.9], [0.9, 0.99]],  # w = True
    ])
    return probs[w.detach(), s.detach(), r.detach()]

In [12]:
p = torch.zeros((2,2,2,2))
for c in range(2):
    for s in range(2):
        for r in range(2):
            for w in range(2):
                p[c, s, r, w] = p_C(torch.tensor(c))*p_S_given_C(torch.tensor(s), torch.tensor(c))*p_R_given_C(torch.tensor(r), torch.tensor(c))*p_W_given_S_R(torch.tensor(w), torch.tensor(s), torch.tensor(r))

cloudy, *sprinker*, rainy, *wet*

In [15]:
p[1,1,1,1]

tensor(0.3240)

In [17]:
p[0,1,0,0] + p[0,1,0,1]

tensor(0.2000)

In [18]:
p

tensor([[[[0.2000, 0.0000],
          [0.0050, 0.0450]],

         [[0.0200, 0.1800],
          [0.0005, 0.0495]]],


        [[[0.0900, 0.0000],
          [0.0360, 0.3240]],

         [[0.0010, 0.0090],
          [0.0004, 0.0396]]]])

In [20]:
def getprob(cloudy, rainy):
    return p[cloudy, 1, rainy, 1]

In [40]:
getprob(0,0)

tensor(0.1800)

In [42]:
getprob(0,1)

tensor(0.0495)

In [43]:
getprob(1,0)

tensor(0.0090)

In [44]:
getprob(1,1)

tensor(0.0396)

In [46]:
getprob(0,1)+getprob(1,1)

tensor(0.0891)

getprob(1,0) but get 0.18, which is getprob(0,0)

0, 1 but get 0.0396, which is 1, 1